warp10/src/main/java/io/warp10/script/ext/http/HTTP.java

```diff
@@ -0,0 +1,442 @@
+    long maxsize;
+    if (null != Capabilities.get(stack, HttpWarpScriptExtension.ATTRIBUTE_HTTP_SIZE)) {
+      maxsize = Long.valueOf(Capabilities.get(stack, HttpWarpScriptExtension.ATTRIBUTE_HTTP_SIZE));
+    } else {
+      maxsize = baseMaxSize;
+    }
+
+    //
+    // Retrieve arguments
+    //
+
+    String method = (String) params.get(METHOD);
+    if (null == method) {
+      throw new WarpScriptException(getName() + " expects an http method.");
+    }
+
+    Map<Object, Object> headers = (Map) params.getOrDefault(HEADERS, new HashMap<>());
+    Object body = params.getOrDefault(BODY,"");
+
+    List authInfo = (List) params.get(AUTH_INFO);
+    WarpScriptStack.Macro authMacro = (WarpScriptStack.Macro) params.get(AUTH_MACRO);
+    Long chunkSize = (Long) params.getOrDefault(CHUNK_SIZE, -1L);
+    WarpScriptStack.Macro chunkMacro = (WarpScriptStack.Macro) params.get(CHUNK_MACRO);
+
+    //
+    // Check URL
+    //
+
+    Object urlParam = params.get(URL);
+    if (null == urlParam) {
+      throw new WarpScriptException(getName() + " expects an url.");
+    }
+
+    URL url = null;
+    try {
+      url = new URL((String) urlParam);
+    } catch (MalformedURLException mue) {
+      throw new WarpScriptException(getName() + " encountered an invalid URL.", mue);
+    }
+
+    if (!"http".equals(url.getProtocol()) && !"https".equals(url.getProtocol())) {
+      throw new WarpScriptException(getName() + " only supports http and https protocols.");
+    }
+
+    if (!webAccessController.checkURL(url)) {
+      throw new WarpScriptException(getName() + " invalid host or scheme in URL.");
+    }
+
+    //
+    // Check that we do not exceed the limits
+    //
+
+    // Get the current counters in the stack and initialize them if not present.
+    AtomicLong urlCount;
+    AtomicLong downloadSize;
+
+    Object ufCount = stack.getAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_REQUESTS);
+    Object ufSize = stack.getAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_SIZE);
+
+    if (null == ufCount || null == ufSize) {
+      urlCount = new AtomicLong();
+      downloadSize = new AtomicLong();
+      stack.setAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_REQUESTS, urlCount);
+      stack.setAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_SIZE, downloadSize);
+    } else {
+      urlCount = (AtomicLong) ufCount;
+      downloadSize = (AtomicLong) ufSize;
+    }
+
+    if (urlCount.addAndGet(1) > maxrequests) {
+      throw new WarpScriptException(getName() + " is limited to " + maxrequests + " calls.");
+    }
+
+    Map<String, Object> res = new HashMap<>();
+    HttpURLConnection conn = null;
+
+    try {
+      conn = (HttpURLConnection) url.openConnection();
+
+      //
+      // Encode userinfo and set headers
+      //
+
+      if (null != authInfo) {
+
+        Map additionalHeaders;
+        if (null != authMacro) {
+          stack.push(authInfo);
+          stack.exec(authMacro);
+          additionalHeaders = (Map) stack.pop();
```

Comment:
I think the macro.auth mechanism will not add any value to the HTTP method. For instance, it will not help in generating the digest auth authorization header, especially when doing multiple authenticated requests.

Reply:
I don't see why it can not be used to generate digest auth. But I see it can be painful to write the auth macro for digest auth without another extension (one could wraps another java function, like for example [DigestScheme.authenticate](https://www.javadoc.io/static/org.apache.httpcomponents/httpclient/4.5.3/org/apache/http/impl/auth/DigestScheme.html#authenticate(org.apache.http.auth.Credentials,%20org.apache.http.HttpRequest,%20org.apache.http.protocol.HttpContext)))..